README
This version of patch contains modifications made by the Free Software
Foundation, summarized in the file ChangeLog.  Primarily they are to
support the unified context diff format that GNU diff can produce, but
they also include some fixes for other bugs.  The FSF is distributing
this version of patch itself because, as of this writing, Larry has
not released a new version of patch since mid-1988.  I have heard that
he has been too busy working on other things, like Perl.

The FSF changes to patch are not covered by the GNU General Public
License; they are distributed on the same terms as the rest of patch
(actually, on somewhat less restrictive terms).

Please send bug reports for this version of patch to
bug-gnu-utils@prep.ai.mit.edu as well as to Larry Wall (lwall@netlabs.com).
 --djm@gnu.ai.mit.edu (David MacKenzie)

			Patch Kit, Version 2.0

		    Copyright (c) 1988, Larry Wall

You may copy the patch kit in whole or in part as long as you don't try to
make money off it, or pretend that you wrote it.
--------------------------------------------------------------------------

Please read all the directions below before you proceed any further, and
then follow them carefully.  Failure to do so may void your warranty. :-)

After you have unpacked your kit, you should have all the files listed
in MANIFEST.

Installation

1)  Run Configure.  This will figure out various things about your system.
    Some things Configure will figure out for itself, other things it will
    ask you about.  It will then proceed to make config.h, config.sh, and
    Makefile.

    You might possibly have to trim # comments from the front of Configure
    if your sh doesn't handle them, but all other # comments will be taken
    care of.

    If you don't have sh, you'll have to rip the prototype of config.h out
    of Configure and generate the defines by hand.

2)  Glance through config.h to make sure system dependencies are correct.
    Most of them should have been taken care of by running the Configure script.

    If you have any additional changes to make to the C definitions, they
    can be done in the Makefile, or in config.h.  Bear in mind that they may
    get undone next time you run Configure.

3)  make

    This will attempt to make patch in the current directory.

4)  make install

    This will put patch into a public directory (normally /usr/local/bin).
    It will also try to put the man pages in a reasonable place.  It will not
    nroff the man page, however.

5)  Read the manual entry before running patch.

6)  IMPORTANT!  Help save the world!  Communicate any problems and
    suggested patches to me, lwall@netlabs.com (Larry Wall),
    so we can keep the world in sync.  If you have a problem, there's
    someone else out there who either has had or will have the same problem.

    If possible, send in patches such that the patch program will apply them.
    Context diffs are the best, then normal diffs.  Don't send ed scripts--
    I've probably changed my copy since the version you have.

    Watch for patch patches in net.sources.bugs.  Patches will generally be
    in a form usable by the patch program.  If you are just now bringing up
    patch and aren't sure how many patches there are, write to me and I'll
    send any you don't have.  Your current patch level is shown in patchlevel.h.


NEW FEATURES IN THIS RELEASE

(Correct) support for 4.3bsd-style context diffs.
Files can be created from scratch.
You can specify a fuzz-factor for context matching.
You can force patch to ask no questions.
You can specify how much of the leading pathname to strip off filenames.
Uses a Configure script for greater portability.
You are now asked if you want to apply a reversed patch.
No limit (apart from memory) on the size of hunks.


ChangeLog
Mon Mar 16 14:10:42 1992  David J. MacKenzie  (djm@wookumz.gnu.ai.mit.edu)

	* patchlevel.h: PATCHLEVEL 12u6.

Sat Mar 14 13:13:29 1992  David J. MacKenzie  (djm at frob.eng.umd.edu)

	* Configure, config.h.SH: Check for directory header and unistd.h.

	* patch.c (main): If -E was given and output file is empty after
	patching, remove it.
	(get_some_switches): Recognize -E option.

	* patch.c (copy_till): Make garbled output an error, not a warning
	that doesn't change the exit status.

	* common.h: Protect against system declarations of malloc and realloc.

	* Makedist: Add backupfile.[ch].

	* Configure: Look for C library where NeXT and SVR4 put it.
	Look in /usr/ucb after /bin and /usr/bin for utilities,
	and look in /usr/ccs/bin, to make SVR4 happier.
	Recognize m68k predefine.

	* util.c (fetchname): Test of stat return value was backward.
	From csss@scheme.cs.ubc.ca.

	* version.c (version): Exit with status 0, not 1.

	* Makefile.SH: Add backupfile.[cho].
	* patch.c (main): Initialize backup file generation.
	(get_some_switches): Add -V option.
	* common.h, util,c, patch.c: Replace origext with simple_backup_suffix.
	* util.c (move_file): Use find_backup_file_name.

Tue Dec  3 11:27:16 1991  David J. MacKenzie  (djm at wookumz.gnu.ai.mit.edu)

	* patchlevel.h: PATCHLEVEL 12u5.

	* Makefile.SH: Change clean, distclean, and realclean targets a
	little so they agree with the GNU coding standards.
	Add Makefile to addedbyconf, so distclean removes it.

	* Configure: Recognize Domain/OS C library in /lib/libc.
	From mmuegel@mot.com (Michael S. Muegel).

	* pch.c: Fixes from Wayne Davison:
	Patch now accepts no-context context diffs that are
	specified with an assumed one line hunk (e.g.  "*** 10 ****").
	Fixed a bug in both context and unified diff processing that would
	put a zero-context hunk in the wrong place (one line too soon).
	Fixed a minor problem with p_max in unified diffs where it would
	set p_max to hunkmax unnecessarily (the only adverse effect was to
	not supply empty lines at eof by assuming they were truncated).

Tue Jul  2 03:25:51 1991  David J. MacKenzie  (djm at geech.gnu.ai.mit.edu)

	* Configure: Check for signal declaration in
	/usr/include/sys/signal.h as well as /usr/include/signal.h.

	* Configure, common.h, config.h.SH: Comment out the sprintf
	declaration and tests to determine its return value type.  It
	conflicts with ANSI C systems' prototypes in stdio.h and the
	return value of sprintf is never used anyway -- it's always cast
	to void.

Thu Jun 27 13:05:32 1991  David J. MacKenzie  (djm at churchy.gnu.ai.mit.edu)

	* patchlevel.h: PATCHLEVEL 12u4.

Thu Feb 21 15:18:14 1991  David J. MacKenzie  (djm at geech.ai.mit.edu)

	* pch.c (another_hunk): Fix off by 1 error.  From
	iverson@xstor.com (Tim Iverson).

Sun Jan 20 20:18:58 1991  David J. MacKenzie  (djm at geech.ai.mit.edu)

	* Makefile.SH (all): Don't make a dummy `all' file.

	* patchlevel.h: PATCHLEVEL 12u3.

	* patch.c (nextarg): New function.
	(get_some_switches): Use it, to prevent dereferencing a null
	pointer if an option that takes an arg is not given one (is last
	on the command line).  From Paul Eggert.

	* pch.c (another_hunk): Fix from Wayne Davison to recognize
	single-line hunks in unified diffs (with a single line number
	instead of a range).

	* inp.c (rev_in_string): Don't use `s' before defining it.  From
	Wayne Davison.

Mon Jan  7 06:25:11 1991  David J. MacKenzie  (djm at geech.ai.mit.edu)

	* patchlevel.h: PATCHLEVEL 12u2.

	* pch.c (intuit_diff_type): Recognize `+++' in diff headers, for
	unified diff format.  From unidiff patch 1.

Mon Dec  3 00:14:25 1990  David J. MacKenzie  (djm at albert.ai.mit.edu)

	* patch.c (get_some_switches): Make the usage message more
	informative.

Sun Dec  2 23:20:18 1990  David J. MacKenzie  (djm at albert.ai.mit.edu)

	* Configure: When checking for C preprocessor, look for 'abc.*xyz'
	instead of 'abc.xyz', so ANSI C preprocessors work.

	* Apply fix for -D from ksb@mentor.cc.purdue.edu (Kevin Braunsdorf).

	* Apply unidiff patches from davison@dri.com (Wayne Davison).

Wed Mar  7 23:47:25 1990  Jim Kingdon  (kingdon at pogo.ai.mit.edu)

	* pch.c: Call malformed instead of goto malformed
	(just allows easier debugging).

Tue Jan 23 21:27:00 1990  Jim Kingdon  (kingdon at pogo.ai.mit.edu)

	* common.h (TMP*NAME): Make these char *, not char [].
	patch.c (main): Use TMPDIR (if present) to set TMP*NAME.
	common.h: Declare getenv.

Sun Dec 17 17:29:48 1989  Jim Kingdon  (kingdon at hobbes.ai.mit.edu)

	* patch.c (reverse_flag_specified): New variable.
	(get_some_switches, reinitialize_almost_everything): Use it.

Local Variables:
version-control: never
end:


inp.c
/* $Header: inp.c,v 2.0.1.1 88/06/03 15:06:13 lwall Locked $
 *
 * $Log:	inp.c,v $
 * Revision 2.0.1.1  88/06/03  15:06:13  lwall
 * patch10: made a little smarter about sccs files
 *
 * Revision 2.0  86/09/17  15:37:02  lwall
 * Baseline for netwide release.
 *
 */

#include "EXTERN.h"
#include "common.h"
#include "util.h"
#include "pch.h"
#include "INTERN.h"
#include "inp.h"

/* Input-file-with-indexable-lines abstract type */

static long i_size;			/* size of the input file */
static char *i_womp;			/* plan a buffer for entire file */
static char **i_ptr;			/* pointers to lines in i_womp */

static int tifd = -1;			/* plan b virtual string array */
static char *tibuf[2];			/* plan b buffers */
static LINENUM tiline[2] = {-1, -1};	/* 1st line in each buffer */
static LINENUM lines_per_buf;		/* how many lines per buffer */
static int tireclen;			/* length of records in tmp file */

/* New patch--prepare to edit another file. */

void
re_input()
{
    if (using_plan_a) {
	i_size = 0;
#ifndef lint
	if (i_ptr != Null(char**))
	    free((char *)i_ptr);
#endif
	if (i_womp != Nullch)
	    free(i_womp);
	i_womp = Nullch;
	i_ptr = Null(char **);
    }
    else {
	using_plan_a = TRUE;		/* maybe the next one is smaller */
	Close(tifd);
	tifd = -1;
	free(tibuf[0]);
	free(tibuf[1]);
	tibuf[0] = tibuf[1] = Nullch;
	tiline[0] = tiline[1] = -1;
	tireclen = 0;
    }
}

/* Constuct the line index, somehow or other. */

void
scan_input(filename)
char *filename;
{
    if (!plan_a(filename))
	plan_b(filename);
    if (verbose) {
	say3("Patching file %s using Plan %s...\n", filename,
	  (using_plan_a ? "A" : "B") );
    }
}

/* Try keeping everything in memory. */

bool
plan_a(filename)
char *filename;
{
    int ifd;
    Reg1 char *s;
    Reg2 LINENUM iline;

    if (ok_to_create_file && stat(filename, &filestat) < 0) {
	if (verbose)
	    say2("(Creating file %s...)\n",filename);
	makedirs(filename, TRUE);
	close(creat(filename, 0666));
    }
    if (stat(filename, &filestat) < 0) {
	Sprintf(buf, "RCS/%s%s", filename, RCSSUFFIX);
	if (stat(buf, &filestat) >= 0 || stat(buf+4, &filestat) >= 0) {
	    Sprintf(buf, CHECKOUT, filename);
	    if (verbose)
		say2("Can't find %s--attempting to check it out from RCS.\n",
		    filename);
	    if (system(buf) || stat(filename, &filestat))
		fatal2("Can't check out %s.\n", filename);
	}
	else {
	    Sprintf(buf+20, "SCCS/%s%s", SCCSPREFIX, filename);
	    if (stat(s=buf+20, &filestat) >= 0 ||
	      stat(s=buf+25, &filestat) >= 0) {
		Sprintf(buf, GET, s);
		if (verbose)
		    say2("Can't find %s--attempting to get it from SCCS.\n",
			filename);
		if (system(buf) || stat(filename, &filestat))
		    fatal2("Can't get %s.\n", filename);
	    }
	    else
		fatal2("Can't find %s.\n", filename);
	}
    }
    filemode = filestat.st_mode;
    if ((filemode & S_IFMT) & ~S_IFREG)
	fatal2("%s is not a normal file--can't patch.\n", filename);
    i_size = filestat.st_size;
#ifdef OS2_16
    if ( i_size > 65500L )
      return FALSE;
#endif
    if (out_of_mem) {
	set_hunkmax();		/* make sure dynamic arrays are allocated */
	out_of_mem = FALSE;
	return FALSE;			/* force plan b because plan a bombed */
    }
#ifdef lint
    i_womp = Nullch;
#else
    i_womp = malloc((MEM)(i_size+2));	/* lint says this may alloc less than */
					/* i_size, but that's okay, I think. */
#endif
    if (i_womp == Nullch)
	return FALSE;
    if ((ifd = open(filename, O_RDONLY|O_BINARY)) < 0)
	fatal2("Can't open file %s\n", filename);
#ifndef lint
    if (read(ifd, i_womp, (int)i_size) != i_size) {
	Close(ifd);	/* probably means i_size > 15 or 16 bits worth */
	free(i_womp);	/* at this point it doesn't matter if i_womp was */
	return FALSE;	/*   undersized. */
    }
#endif
    Close(ifd);
    if (i_size && i_womp[i_size-1] != '\n')
	i_womp[i_size++] = '\n';
    i_womp[i_size] = '\0';

    /* count the lines in the buffer so we know how many pointers we need */

    iline = 0;
    for (s=i_womp; *s; s++) {
	if (*s == '\n')
	    iline++;
    }
#ifdef lint
    i_ptr = Null(char**);
#else
    i_ptr = (char **)malloc((MEM)((iline + 2) * sizeof(char *)));
#endif
    if (i_ptr == Null(char **)) {	/* shucks, it was a near thing */
	free((char *)i_womp);
	return FALSE;
    }

    /* now scan the buffer and build pointer array */

    iline = 1;
    i_ptr[iline] = i_womp;
    for (s=i_womp; *s; s++) {
	if (*s == '\n')
	    i_ptr[++iline] = s+1;	/* these are NOT null terminated */
    }
    input_lines = iline - 1;

    /* now check for revision, if any */

    if (revision != Nullch) {
	if (!rev_in_string(i_womp)) {
	    if (force) {
		if (verbose)
		    say2(
"Warning: this file doesn't appear to be the %s version--patching anyway.\n",
			revision);
	    }
	    else {
		ask2(
"This file doesn't appear to be the %s version--patch anyway? [n] ",
		    revision);
	    if (*buf != 'y')
		fatal1("Aborted.\n");
	    }
	}
	else if (verbose)
	    say2("Good.  This file appears to be the %s version.\n",
		revision);
    }
    return TRUE;			/* plan a will work */
}

/* Keep (virtually) nothing in memory. */

void
plan_b(filename)
char *filename;
{
    Reg3 FILE *ifp;
    Reg1 int i = 0;
    Reg2 int maxlen = 1;
    Reg4 bool found_revision = (revision == Nullch);

    using_plan_a = FALSE;
    if ((ifp = fopen(filename, "rb")) == Nullfp)
	fatal2("Can't open file %s\n", filename);
    if ((tifd = open(TMPINNAME, O_WRONLY|O_CREAT|O_TRUNC|O_BINARY,
                     S_IWRITE|S_IREAD)) < 0)
	fatal2("Can't open file %s\n", TMPINNAME);
    while (fgets(buf, sizeof buf, ifp) != Nullch) {
	if (revision != Nullch && !found_revision && rev_in_string(buf))
	    found_revision = TRUE;
	if ((i = strlen(buf)) > maxlen)
	    maxlen = i;			/* find longest line */
    }
    if (revision != Nullch) {
	if (!found_revision) {
	    if (force) {
		if (verbose)
		    say2(
"Warning: this file doesn't appear to be the %s version--patching anyway.\n",
			revision);
	    }
	    else {
		ask2(
"This file doesn't appear to be the %s version--patch anyway? [n] ",
		    revision);
		if (*buf != 'y')
		    fatal1("Aborted.\n");
	    }
	}
	else if (verbose)
	    say2("Good.  This file appears to be the %s version.\n",
		revision);
    }
    rewind(ifp);          /* rewind file */
    lines_per_buf = BUFFERSIZE / maxlen;
    tireclen = maxlen;
    tibuf[0] = malloc((MEM)(BUFFERSIZE + 1));
    tibuf[1] = malloc((MEM)(BUFFERSIZE + 1));
    if (tibuf[1] == Nullch)
	fatal1("Can't seem to get enough memory.\n");
    for (i=1; ; i++) {
	if (! (i % lines_per_buf))	/* new block */
	    if (write(tifd, tibuf[0], BUFFERSIZE) < BUFFERSIZE)
		fatal1("patch: can't write temp file.\n");
	if (fgets(tibuf[0] + maxlen * (i%lines_per_buf), maxlen + 1, ifp)
	  == Nullch) {
	    input_lines = i - 1;
	    if (i % lines_per_buf)
		if (write(tifd, tibuf[0], BUFFERSIZE) < BUFFERSIZE)
		    fatal1("patch: can't write temp file.\n");
	    break;
	}
    }
    Fclose(ifp);
    Close(tifd);

    if ((tifd = open(TMPINNAME, O_RDONLY|O_BINARY)) < 0) {
	fatal2("Can't reopen file %s\n", TMPINNAME);
    }
}

/* Fetch a line from the input file, \n terminated, not necessarily \0. */

char *
ifetch(line,whichbuf)
Reg1 LINENUM line;
int whichbuf;				/* ignored when file in memory */
{
    if (line < 1 || line > input_lines)
	return "";
    if (using_plan_a)
	return i_ptr[line];
    else {
	LINENUM offline = line % lines_per_buf;
	LINENUM baseline = line - offline;

	if (tiline[0] == baseline)
	    whichbuf = 0;
	else if (tiline[1] == baseline)
	    whichbuf = 1;
	else {
	    tiline[whichbuf] = baseline;
#ifndef lint		/* complains of long accuracy */
	    Lseek(tifd, (long)baseline / lines_per_buf * BUFFERSIZE, 0);
#endif
	    if (read(tifd, tibuf[whichbuf], BUFFERSIZE) < 0)
		fatal2("Error reading tmp file %s.\n", TMPINNAME);
	}
	return tibuf[whichbuf] + (tireclen*offline);
    }
}

/* True if the string argument contains the revision number we want. */

bool
rev_in_string(string)
char *string;
{
    Reg1 char *s;
    Reg2 int patlen;

    if (revision == Nullch)
	return TRUE;
    patlen = strlen(revision);
    if (strnEQ(string,revision,patlen) && isspace(string[patlen]))
	return TRUE;
    for (s = string; *s; s++) {
	if (isspace(*s) && strnEQ(s+1, revision, patlen) &&
		isspace(s[patlen+1] )) {
	    return TRUE;
	}
    }
    return FALSE;
}

